Libraries/LibWeb/Loader/ResourceLoader.cpp

```diff
@@ -184,6 +184,7 @@ void ResourceLoader::load(const LoadRequest& request, Function<void(ReadonlyByte
                     error_callback("HTTP load failed");
                 return;
             }
+            deferred_invoke([&](auto&) { const_cast<RefPtr<Protocol::Download>&>(download) = nullptr; });
```

Comment:
Wait, what's going on here?

`deferred_invoke` will call this lambda when returning to the event loop. By that point, won't we be somewhere else entirely, and not have `download` on the stack anymore?

Reply:
`download` is captured as a copy (of a `RefPtr`)
the lambda keeps the download alive until success_callback (or the error one) is called and done

it's such a convoluted mess, but I can't find a better way that doesn't involve making LibWeb support streaming downloads

Reply:
Oh.. okay.

Reply:
precisely my reaction when I realised _that_ was why the downloads weren't being deleted after they finished

Reply:
@alimpfard `download` is certainly not captured as a copy because of `[&]`! It probably works because nothing on the stack has trashed it, but this is definitely merely a reference capture to a local variable.

Reply:
This line looks really weird and looks like it will trash the stack at that location. This line should not be necessary, and the reason why it doesn't get freed is probably because `download` is copy-captured here:

https://github.com/alimpfard/serenity/blob/b87e92d70654509d69511c4923f2ca787f003d06/Libraries/LibWeb/Loader/ResourceLoader.cpp#L173

So, unless something resets `download->on_buffered_download_finish` you end up with a circular reference. So, I think all you really need is doing this instead to clear the circular reference:
```c++
deferred_invoke([download](auto&) { download->on_buffered_download_finish = nullptr; });
```

Reply:
The `success_callback` may keep the underlying download object alive, but not necessarily the RefPtr pointing to it.

Reply:
Hmmm...
Here's the way I see that:
`download` is a RefPtr, and LibProtocol keeps it alive until it calls `on_finish`
`download->on_buffered_download_finish` is some object (internal to `Function`) on the heap.
giving a copy of `download` to `on_buffered_download_finish` creates a cyclic reference chain between `download` and `on_buffered_download_finish`, where `download` will end up with a refcount of 1.

taking a reference to `download` in the `deferred_invoke` lambda should be fine, as it's not pointing into stack - it's pointing to something kept around by `on_buffered_download_finish` (which is kept alive by `download` and is keeping `download` alive).
Clearing `download` will drop its refcount, which will in turn destroy `on_buffered_download_finished` and break the cycle.
it's too complicated and weird though.

where did that `success_callback` keeping download alive come from though?

Reply:
You're right, the reference probably points to the copy-captured value of the other lambda. So this code assumes that `on_buffered_download_finish` doesn't get cleared before the lambda referencing `download` finishes execution. Seems a little brittle but I guess works...